server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -190,7 +233,9 @@ private Samples getSamples(List<Reference> columns,
                     Engine.Searcher searcher = indexShard.acquireSearcher("update-table-statistics");
                     searchersToRelease.add(searcher);
                     totalNumDocs += searcher.getIndexReader().numDocs();
-                    totalSizeInBytes += indexShard.storeStats().getSizeInBytes();
+                    long bytesRead = indexShard.storeStats().getSizeInBytes();
+                    totalSizeInBytes += bytesRead;
+                    maybePause(bytesRead);
```

Comment:
I think most of the time will be spent in the `ReservoirCollector` and further below where we load the values from disk: 

https://github.com/crate/crate/blob/a3eb7f078f1e7981b7489e03ecf93d027bb575b0/server/src/main/java/io/crate/statistics/ReservoirSampler.java#L214-L215

Only throttling up here is probably not having a large impact.

Reply:
To elaborate on this - if we do the check based on the size of the indexShard we'll have _one_ large pause per shard and still cause large IO spikes. What we want is to smooth out the IO access.

For that we'd need to account bytes read *per row* and pause during the inner loop.

Reply:
thx for the details, done in https://github.com/crate/crate/pull/13046/commits/a25da8cbe3f3c6dfb74fc12f2a32c94149426c3e

Reply:
commit ^ had a bug - I acidentally added row size to `totalSizeInBytes`. In the master branch we add only `indexShard` stats size.

 Pushed a fixup https://github.com/crate/crate/pull/13046/commits/1daf8d8cd6320e299e2f9afacbb8ddd938fadb2a but it's a bit worrying that all tests passed even with buggy version. 
 
Looks like `totalSizeInBytes` computed in `ReservoirSampler` is not used in `pg_stats` and only used in some optimizers - so it's probably fine that there is no visible effect of that bug.
 
 `Samples.totalSizeInBytes` is used in `Stats.sizeInBytes` which is indirectly used **only** in `Stats.averageSizePerRowInBytes` (`Stats.sizeInBytes` is unused directly somewhere else) ==> this is used in couple of optimisers. I cleaned up unused code in https://github.com/crate/crate/pull/13046/commits/59cab831ca4dc56f6eb23e74c0ebb80ff3f4dba1